stan/math/rev/core/print_stack.hpp

```diff
@@ -22,11 +22,8 @@ inline void print_stack(std::ostream& o) {
     << std::endl;
   // TODO(carpenter): this shouldn't need to be cast any more
   for (size_t i = 0; i < ChainableStack::instance_->var_stack_.size(); ++i) {
-    o << i << "  " << ChainableStack::instance_->var_stack_[i] << "  "
-      << (static_cast<vari*>(ChainableStack::instance_->var_stack_[i]))->val_
-      << " : "
-      << (static_cast<vari*>(ChainableStack::instance_->var_stack_[i]))->adj_
-      << std::endl;
+    boost::apply_visitor(vari_printer(o, i),
```

Comment:
Given this doesn't need to be fast, we can just lean on polymorphism here right? The code does look nicer that way. [This](https://www.bfilipek.com/2020/04/variant-virtual-polymorphism.html#performance) page says with variants you can get away from doing polymorphism at all. Is that something desirable at all? I think just doing variants on the `grad` and `set_zero_all_adjoint` loops handles the bits of the code where polymorphism is slow.

Reply:
> Given this doesn't need to be fast, we can just lean on polymorphism here right?

`vari_base` doesn't have the concept of `adj_` and `val_` so idt you can actually use polymorphism here (and it doesn't exist anymore).

>   This page says with variants you can get away from doing polymorphism at all. Is that something desirable at all? I think just doing variants on the grad and set_zero_all_adjoint loops handles the bits of the code where polymorphism is slow.

So yes if we wanted to we could have that `boost::variant` actually hold all of the possible vari structs. But that's a lot! We'd have to have some sort of macro that would let us register a vari type that can go into the boost variant. It's an interesting idea but I think maintenance would be an issue

Reply:
> But that's a lot! We'd have to have some sort of macro

Well if we don't have a vari_base type then we already did this?

Reply:
Nope! `vari_variant` is

```cpp
using vari_variant = boost::variant<vari_value<double>*, vari_value<float>*,
                                    vari_value<long double>*>;
```

So when a struct like `op_vd_vari` inherits from `vari_value<double>` polymorphism allows that `op_vd_vari` to fit in the `vari_value<double>`. So instead of having to list all of them explicitly, we just have to list out the base vari's they derive from

Reply:
Oooh, okay, makes sense.

Reply:
Is that not very wasteful with memory? The linked post says variant is of the size of the largest type. I assume we will still mostly use `vari_value<double>`, but this way each will be of the size of `vari_value<long double>`. 

I don't think you intend to implement mixed type autodiff just yet. So maybe you can only focus on double here or have some global typedef/macro for chosing the scalar floating point type.

Reply:
Ah! I was worried about the same thing but they are all pointers!